### PR TITLE
Http-server

### PR DESCRIPTION
**Pull request by benjamin-wilson**
Rest server, file server, website, OTA Updates, AP mode

### CODE REVIEW

**Comment by benjamin-wilson** [comment]
Will get squashed and merged

File: CMakeLists.txt [hunk 1]
```diff
@@ -6,5 +6,12 @@ cmake_minimum_required(VERSION 3.5)
 # This example uses an extra component for common functions such as Wi-Fi and Ethernet connection.
 # set(EXTRA_COMPONENT_DIRS $ENV{IDF_PATH}/examples/common_components/protocol_examples_common)
 
+set(SDKCONFIG "${CMAKE_BINARY_DIR}/sdkconfig")
```
Comment (johnny9) [inline CMakeLists.txt:9]: Why move the SDKCONFIG path from the default?

File: CMakeLists.txt [hunk 2]
```diff
@@ -6,5 +6,12 @@ cmake_minimum_required(VERSION 3.5)
...
+set(SDKCONFIG_DEFAULTS "sdkconfig.defaults")
```
Comment (johnny9) [inline CMakeLists.txt:10]: Think this is redundant as sdkconfig.defaults is already the normal default config

File: main/nvs_config.c [hunk 3]
```diff
@@ -40,6 +40,29 @@ char * nvs_config_get_string(const char * key, const char * default_value)
     return out;
 }
 
+void nvs_config_set_string(const char * key, const char * value){
```
Comment (johnny9) [inline main/nvs_config.c:43]:
```suggestion
void nvs_config_set_string(const char * key, const char * value)
{
```

File: main/nvs_config.c [hunk 4]
```diff
@@ -40,6 +40,29 @@ char * nvs_config_get_string(const char * key, const char * default_value)
...
+    err = nvs_set_str(handle, key, value);
+    if (err != ESP_OK)
+    {
+        ESP_LOGW(TAG, "Could not write nvs key: %s, value: %s", key, value);
+        return;
+    }
+
+    nvs_close(handle);
+    return;
+
```
Comment (johnny9) [inline main/nvs_config.c:63]:
```suggestion
```

File: main/nvs_config.c [hunk 5]
```diff
@@ -63,3 +86,27 @@ uint16_t nvs_config_get_u16(const char * key, const uint16_t default_value)
     nvs_close(handle);
     return out;
 }
+
+
+void nvs_config_set_u16(const char * key, const uint16_t value){
+
```
Comment (johnny9) [inline main/nvs_config.c:92]:
```suggestion
```

File: main/nvs_config.c [hunk 6]
```diff
@@ -63,3 +86,27 @@ uint16_t nvs_config_get_u16(const char * key, const uint16_t default_value)
     nvs_close(handle);
     return out;
 }
+
+
+void nvs_config_set_u16(const char * key, const uint16_t value){
```
Comment (johnny9) [inline main/nvs_config.c:91]:
```suggestion
void nvs_config_set_u16(const char * key, const uint16_t value)
{
```

File: main/nvs_config.c [hunk 7]
```diff
@@ -63,3 +86,27 @@ uint16_t nvs_config_get_u16(const char * key, const uint16_t default_value)
...
+    err = nvs_set_u16(handle, key, value);
+    if (err != ESP_OK)
+    {
+        ESP_LOGW(TAG, "Could not write nvs key: %s, value: %u", key, value);
+        return;
+    }
+
+    nvs_close(handle);
+    return;
+
```
Comment (johnny9) [inline main/nvs_config.c:111]:
```suggestion
```

File: main/tasks/power_management_task.c [hunk 8]
```diff
@@ -54,6 +54,7 @@ void POWER_MANAGEMENT_task(void * pvParameters){
         power_management->fan_speed = EMC2101_get_fan_speed();
 
         if(strcmp(ASIC_MODEL, "BM1397") == 0){
+
```
Comment (johnny9) [inline main/tasks/power_management_task.c:57]: dont add extra lines

File: main/tasks/user_input_task.c [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+void USER_INPUT_task(void * pvParameters){
+
+    gpio_set_direction(BUTTON_BOOT, GPIO_MODE_INPUT);
+
+    while(1) {
+        if (gpio_get_level(BUTTON_BOOT) == 0 && button_being_pressed == false) {  // If button is pressed
+            ESP_LOGI(TAG, "BUTTON PRESSED");
+            button_being_pressed = true;
+            toggle_wifi_softap();
+        }else if(gpio_get_level(BUTTON_BOOT) == 1 && button_being_pressed == true){
```
Comment (johnny9) [inline main/tasks/user_input_task.c:21]:
```suggestion
        } else if (gpio_get_level(BUTTON_BOOT) == 1 && button_being_pressed == true) {
```

File: main/http_server/http_server.c [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+    httpd_resp_sendstr(req, "WWW update complete\n");
+    return ESP_OK;
+}
+
+/*
+ * Handle OTA file upload
+ */
+esp_err_t POST_OTA_update(httpd_req_t *req)
+{
+	char buf[1000];
```
Comment (johnny9) [inline main/http_server/http_server.c:300]: This function appears to be all tabs for indentations

File: main/http_server/http_server.c [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+
+		// Serious Error: Abort OTA
+		} else if (recv_len <= 0) {
+			httpd_resp_send_err(req, HTTPD_500_INTERNAL_SERVER_ERROR, "Protocol Error");
+			return ESP_FAIL;
+		}
+
+		// Successful Upload: Flash firmware chunk
+		if (esp_ota_write(ota_handle, (const void *)buf, recv_len) != ESP_OK) {
+			httpd_resp_send_err(req, HTTPD_500_INTERNAL_SERVER_ERROR, "Flash Error");
```
Comment (johnny9) [inline main/http_server/http_server.c:322]: might need to call  esp_ota_abort(ota_handle); if there is a failure

File: main/http_server/http_server.c [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+	vTaskDelay(500 / portTICK_PERIOD_MS);
+	esp_restart();
+
+	return ESP_OK;
+}
+
+
+esp_err_t start_rest_server(void *pvParameters)
+{
+    //ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_APSTA));
```
Comment (johnny9) [inline main/http_server/http_server.c:346]: delete if not needed

File: main/http_server/http_server.c [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+static esp_err_t POST_restart(httpd_req_t *req)
+{
+    esp_restart();
+    return ESP_OK;
+}
+
+/* Simple handler for getting system handler */
+static esp_err_t GET_system_info(httpd_req_t *req)
+{
+
```
Comment (johnny9) [inline main/http_server/http_server.c:223]:
```suggestion
```

File: main/http_server/http_server.c [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+    if (ret != ESP_OK) {
+        ESP_LOGE(TAG, "Failed to get SPIFFS partition information (%s)", esp_err_to_name(ret));
+    } else {
+        ESP_LOGI(TAG, "Partition size: total: %d, used: %d", total, used);
+    }
+    return ESP_OK;
+}
+
+
+
```
Comment (johnny9) [inline main/http_server/http_server.c:90]: remove extra lines

File: main/http_server/axe-os/.vscode/settings.json [hunk 15]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (johnny9) [inline main/http_server/axe-os/.vscode/settings.json:1]: Should this be checked in. What is cSpell?

**Review by johnny9 — COMMENTED** [review]
Overall changes look really good. Commented on a few small things.

File: components/connect/connect.c [hunk 16]
```diff
@@ -77,23 +77,52 @@ static void event_handler(void* arg, esp_event_base_t event_base, int32_t event_
...
+}
+
+void toggle_wifi_softap(void){
+    wifi_mode_t mode = WIFI_MODE_NULL;
+    ESP_ERROR_CHECK(esp_wifi_get_mode(&mode));
+
+    if(mode == WIFI_MODE_APSTA){
+        ESP_LOGI(TAG, "ESP_WIFI Access Point Off");
+        ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+    }else{
```
Comment (johnny9) [inline components/connect/connect.c:112]:
```suggestion
    } else {
```

File: components/connect/connect.c [hunk 17]
```diff
@@ -77,23 +77,52 @@ static void event_handler(void* arg, esp_event_base_t event_base, int32_t event_
...
 
-    wifi_config_t wifi_config = {
+    return esp_netif_ap;
+}
+
+void toggle_wifi_softap(void){
+    wifi_mode_t mode = WIFI_MODE_NULL;
+    ESP_ERROR_CHECK(esp_wifi_get_mode(&mode));
+
+    if(mode == WIFI_MODE_APSTA){
```
Comment (johnny9) [inline components/connect/connect.c:109]:
```suggestion
    if (mode == WIFI_MODE_APSTA) {
```

File: components/connect/connect.c [hunk 18]
```diff
@@ -77,23 +77,52 @@ static void event_handler(void* arg, esp_event_base_t event_base, int32_t event_
...
+        ESP_LOGI(TAG, "ESP_WIFI Access Point Off");
+        ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+    }else{
+        ESP_LOGI(TAG, "ESP_WIFI Access Point On");
+        ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_APSTA));
+
```
Comment (johnny9) [inline components/connect/connect.c:115]:
```suggestion
```

File: main/http_server/axe-os/.vscode/settings.json (hunk 15 above)
Comment (benjamin-wilson) [reply main/http_server/axe-os/.vscode/settings.json:1]: Yes, this is so VSCode settings are the same for this project. cSpell is spelling and grammar extension, it'll pick up on words like toastr which are valid words.

File: CMakeLists.txt (hunk 2 above)
Comment (benjamin-wilson) [reply CMakeLists.txt:10]: We need overrides such as CONFIG_ESPTOOLPY_FLASHSIZE="8MB", this is not the default flash size

File: CMakeLists.txt (hunk 1 above)
Comment (benjamin-wilson) [reply CMakeLists.txt:9]: For whatever reason if I remove this line or point it to the root directory the sdkconfig.defaults don't work.

**Pull request merged**